src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/NamingStyle.cs

```diff
@@ -0,0 +1,289 @@
+            {
+                case Capitalization.PascalCase:
+                    if (words.All(w => char.IsUpper(w[0])))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        var violations = words.All(w => !char.IsUpper(w[0]));
+                        failureReason = string.Format(FeaturesResources.NamingViolation_WordsMustBeginWithUppercase, string.Join(", ", violations));
+                        return false;
+                    }
+                case Capitalization.CamelCase:
+                    if (char.IsLower(words.First()[0]) && words.Skip(1).All(w => char.IsUpper(w[0])))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        if (!char.IsLower(words.First()[0]))
+                        {
+                            failureReason = FeaturesResources.NamingViolation_FirstWordMustBeginWithLowercase;
+                        }
+
+                        var violations = words.Skip(1).Where(w => !char.IsUpper(w[0]));
+                        if (violations.Any())
+                        {
+                            if (failureReason != string.Empty)
+                            {
+                                failureReason += Environment.NewLine;
+                            }
+
+                            failureReason += string.Format(FeaturesResources.NamingViolation_NonLeadingWordsMustBeginWithUppercase, string.Join(", ", violations));
+                        }
+
+                        return false;
+                    }
+                case Capitalization.FirstUpper:
+                    if (char.IsUpper(words.First()[0]) && words.Skip(1).All(w => char.IsLower(w[0])))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        if (!char.IsUpper(words.First()[0]))
+                        {
+                            failureReason = string.Format(FeaturesResources.NamingViolation_FirstWordMustBeginWithUppercase, words.First());
+                        }
+
+                        var violations = words.Skip(1).Where(w => !char.IsLower(w[0]));
+                        if (violations.Any())
+                        {
+                            if (failureReason != string.Empty)
+                            {
+                                failureReason += Environment.NewLine;
+                            }
+
+                            failureReason += string.Format(FeaturesResources.NamingViolation_NonLeadingWordsMustBeginWithLowercase, string.Join(", ", violations));
+                        }
+
+                        return false;
+                    }
+                case Capitalization.AllUpper:
+                    if (words.SelectMany(w => w.ToCharArray()).All(c => char.IsUpper(c)))
```

Comment:
Glob question maybe @davkean can help with: Will `char.IsUpper` behave as expected for things like surrogate pairs, etc?